extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaConsumerMonitor.java

```diff
@@ -0,0 +1,91 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.kafka;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.apache.druid.java.util.emitter.service.ServiceEmitter;
+import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
+import org.apache.druid.java.util.metrics.AbstractMonitor;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class KafkaConsumerMonitor extends AbstractMonitor
+{
+  private volatile boolean stopAfterNext = false;
+
+  // Kafka metric name -> Druid metric name
+  private static final Map<String, String> METRICS =
+      ImmutableMap.<String, String>builder()
+                  .put("bytes-consumed-total", "kafka/consumer/bytesConsumed")
+                  .put("records-consumed-total", "kafka/consumer/recordsConsumed")
+                  .build();
+  private static final String TOPIC_TAG = "topic";
+  private static final Set<String> TOPIC_METRIC_TAGS = ImmutableSet.of("client-id", TOPIC_TAG);
+
+  private final KafkaConsumer<?, ?> consumer;
+  private final AtomicLong currentBytesRead = new AtomicLong();
+
+  public KafkaConsumerMonitor(final KafkaConsumer<?, ?> consumer)
+  {
+    this.consumer = consumer;
+  }
+
+  @Override
+  public boolean doMonitor(final ServiceEmitter emitter)
+  {
+    for (final Map.Entry<MetricName, ? extends Metric> entry : consumer.metrics().entrySet()) {
+      final MetricName metricName = entry.getKey();
+
+      if (METRICS.containsKey(metricName.name()) && isTopicMetric(metricName)) {
+        final String topic = metricName.tags().get(TOPIC_TAG);
+        final long newValue = ((Number) entry.getValue().metricValue()).longValue();
+        final long priorValue = currentBytesRead.getAndSet(newValue);
+
+        if (newValue != priorValue) {
+          final ServiceMetricEvent.Builder builder =
+              new ServiceMetricEvent.Builder().setDimension(TOPIC_TAG, topic);
+          emitter.emit(builder.build(METRICS.get(metricName.name()), newValue - priorValue));
+        }
+      }
+    }
+
+    return !stopAfterNext;
+  }
+
+  public void stopAfterNextEmit()
+  {
+    stopAfterNext = true;
+  }
+
+  private static boolean isTopicMetric(final MetricName metricName)
+  {
+    // Certain metrics are emitted both as grand totals and broken down by topic; we want to ignore the grand total and
+    // only look at the per-topic metrics. See https://kafka.apache.org/documentation/#new_consumer_fetch_monitoring.
```

Comment:
Ah, yes, I'm not sure where I got the `new_` from. Fixed.

> since we only talk to one topic i assume there shouldn't be any difference between the grand total and per topic metrics?

I assume so too, although I did it this way to future-proof against a possible future scenario where we support reading from multiple topics. It's also nice to have `topic` as an attribute for the metric.